File: java/src/edu/berkeley/cs/sort/Shell.java
package edu.berkeley.cs.sort;

public class Shell<T extends Comparable> extends Sort<T> {
  @Override
  public void sort(T[] input) {
    int h = 1;
    while (h < input.length / 3) {
      h = 3 * h + 1;
    }

    while (h >= 1) {
      for (int i = h; i < input.length; i++) {
        for (int j = i; j >= h && less(input[j], input[j - h]); j -= h) {
          swap(input, j, j - h);
        }
      }

      h = h / 3;
    }
  }
}

File: python/src/sort/shell.py
from python.src.sort.sort import Sort


class Shell(Sort):

    def sort(self, input):
        h = 1
        while h < len(input) / 3:
            h = int(3 * h + 1)

        while h >= 1:
            for i in range(h, len(input)):
                j = i
                while j >= h and self.less(input[j], input[j - h]):
                    self.swap(input, j, j - h)
                    j -= h

            h = int(h / 3)
